Maps and Users APIs

omermadmon opened the pull request:
Done:
1. Fixed servlets’ documentation.
2. Turned achievements management from hard-coded into servlet-based.
3. Displayed achievements on a map (\achievements.html) using Google Maps API.
4. Added authentication to associate users’ comments with email using Users API.
   * This association is not (yet) visible to the user - it only adds to the achievement entity the user's email address as a property (if logged in). This change has been tested via datastore dashboard.
 
To Do:
1. Implement Comment and Achievement objects using Protocol Buffers - 
   * Currently trying to download the protobuffs compiler (protoc) and use it in appengine (I might ask for assistance if it won’t be solved soon).
2. Optimize random achievement retrieval (RandomAchievementServlet.java) - 
   * Figure out what can be done instead of retrieving all entities and only then generate a random entity to write to the response.
3. Custom comments page when logged in - 
   * Default name in comments form.
   * Enable delete button only for current user’s comments).
 

Webapp live server url: https://omermadmon-step-2020.ew.r.appspot.com/

nolaviz commented on portfolio/src/main/java/com/google/sps/servlets/RandomAchievementServlet.java:
```diff
@@ -0,0 +1,73 @@
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    // Store all achievements in a list.
```
This may use up a lot of memory, if the comments start growing.
Consider instead using results.countEntities() to determine the collection size, then choose which comment to retrieve - and ignore all others.

omermadmon replied:
> This may use up a lot of memory, if the comments start growing.
> Consider instead using results.countEntities() to determine the collection size, then choose which comment to retrieve - and ignore all others.

Been looking for something just like that!
Thanks!

omermadmon commented:
Fixed by applying the following changes:

1. Added a serial number property to achievement entities in datastore (hard coded). This property is used only for achievement randomization
2. Altered the random achievement servlet (including: dealing with the case when there are no achievements at all, and the random achievement query return null).
3. Altered the new achievement servlet to increase the serial by one every time a new achievement is added.

Note: I did not alter the Achievement class, because serial property should be encapsulated (UI should not interact with this property because its only role is for achievement randomization, which is held by the servlet).

nolaviz commented on portfolio/src/main/java/com/google/sps/servlets/NewAchievementServlet.java:
```diff
@@ -29,16 +34,24 @@
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Maybe a nit, but - I just noticed that you always create a new instance of DatastoreService just-in-time (within the methods), here as well as in all the code already written.
From looking at examples / tutorials in the AppEngine documentation, it seems the more common usage is to have a single class-wide DatastoreService field, initialized just once at class-loading time. (In cases where you have a really complex / high-load server, this makes a difference: initializing a service connection can be costly.)

omermadmon replied:
> Maybe a nit, but - I just noticed that you always create a new instance of DatastoreService just-in-time (within the methods), here as well as in all the code already written.
> From looking at examples / tutorials in the AppEngine documentation, it seems the more common usage is to have a single class-wide DatastoreService field, initialized just once at class-loading time. (In cases where you have a really complex / high-load server, this makes a difference: initializing a service connection can be costly.)

Sounds reasonable, will be fixed. 
Thanks.

nolaviz commented on portfolio/src/main/java/com/google/sps/servlets/NewAchievementServlet.java:
```diff
@@ -29,16 +34,24 @@
     // Create new achievement entity.
     Entity achievementEntity = new Entity("Achievement");
+    int serial = results.countEntities();
+    String achievementText = request.getParameter("text") + "\nSerial number: " + serial;
```
Why include the serial number in the text? I thought you claimed it's an internal detail, and not interesting to the end-users?
Also - if you *do* want to present the serial number to the user, instead of storing the serial number twice (both here and in the dedicated "serial" property), prefer to just fetch the "serial" when reading. (We prefer not to modify user-provided textual data, as the results aren't always as expected in some edge-cases.)

omermadmon replied:
> Why include the serial number in the text? I thought you claimed it's an internal detail, and not interesting to the end-users?
> Also - if you _do_ want to present the serial number to the user, instead of storing the serial number twice (both here and in the dedicated "serial" property), prefer to just fetch the "serial" when reading. (We prefer not to modify user-provided textual data, as the results aren't always as expected in some edge-cases.)

Actually, this class is used only for testing and isn't accessible from the UI at all. I use it whenever I run the webapp on a localhost for having some dummy data.

The reason I attached the serial number to the achievement text is for self-validation. I will refactor it.

nolaviz commented on portfolio/src/main/java/com/google/sps/servlets/RandomAchievementServlet.java:
```diff
@@ -39,35 +41,49 @@
 
-    // Generate random achievement.
-    Achievement randomAchievement = achievements.get((int) (Math.random() * achievements.size())); 
+    if(entity != null) {
```
Style nit: missing space after "if".

omermadmon replied:
> Style nit: missing space after "if".

Fixed.

nolaviz commented on portfolio/src/main/java/com/google/sps/servlets/RandomAchievementServlet.java:
```diff
@@ -39,35 +41,49 @@
-    // Generate random achievement.
-    Achievement randomAchievement = achievements.get((int) (Math.random() * achievements.size())); 
+    if(entity != null) {
+        // Create an achievement from the random entity.
```
I recall seeing identical code in ListAchievementsServlet.
Consider refactoring (maybe just add a constructor to create an Achievement from an Entity?).

omermadmon replied:
> I recall seeing identical code in ListAchievementsServlet.
> Consider refactoring (maybe just add a constructor to create an Achievement from an Entity?).

At this servlet, if there are no 'Achievement' entities at all in the datastore, the query returns 'null', so a Achievement object cannot be created from the null we've got instead of the desired entity. 

In the ListAchievementsServlet we just want to retrieve all existing achievement entities, so there is no need to check if some query result is null (in such case, a null json will be written to the response and that fine because there are no achievements to display).

What am I missing here?

nolaviz commented on portfolio/src/main/webapp/script.js:
```diff
@@ -20,15 +20,22 @@ function generateRandomAchievement() {
-        const club = achievement.club;
-        const text = achievement.text;
-        const imgUrl = 'images/' + club + '-logo.png';
+        if(achievement == null) {
```
Nit: missing space after "if".
